remote/remote_state_gcs.go

```diff
@@ -0,0 +1,405 @@
+	if config.GCSBucketLabels == nil || len(config.GCSBucketLabels) == 0 {
+		terragruntOptions.Logger.Printf("No labels specified for bucket %s.", config.remoteStateConfigGCS.Bucket)
+		return nil
+	}
+
+	terragruntOptions.Logger.Printf("Adding labels to GCS bucket with %s", config.GCSBucketLabels)
+
+	ctx := context.Background()
+	bucket := gcsClient.Bucket(config.remoteStateConfigGCS.Bucket)
+
+	bucketAttrs := *&storage.BucketAttrsToUpdate{}
+
+	for key, value := range config.GCSBucketLabels {
+		bucketAttrs.SetLabel(key, value)
+	}
+
+	_, err := bucket.Update(ctx, bucketAttrs)
+
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	return nil
+
+}
+
+// Create the GCS bucket specified in the given config
+func CreateGCSBucket(gcsClient *storage.Client, config *ExtendedRemoteStateConfigGCS, terragruntOptions *options.TerragruntOptions) error {
+	terragruntOptions.Logger.Printf("Creating GCS bucket %s in project %s", config.remoteStateConfigGCS.Bucket, config.Project)
+
+	// The project ID to which the bucket belongs. This is only used when creating a new bucket during initialization.
+	// Since buckets have globally unique names, the project ID is not required to access the bucket during normal
+	// operation.
+	projectID := config.Project
+
+	ctx := context.Background()
+	bucket := gcsClient.Bucket(config.remoteStateConfigGCS.Bucket)
+
+	bucketAttrs := &storage.BucketAttrs{}
+
+	if config.Location != "" {
+		terragruntOptions.Logger.Printf("Creating GCS bucket in location %s.", config.Location)
+	}
+
+	if config.SkipBucketVersioning {
+		terragruntOptions.Logger.Printf("Versioning is disabled for the remote state GCS bucket %s using 'skip_bucket_versioning' config.", config.remoteStateConfigGCS.Bucket)
+	} else {
+		terragruntOptions.Logger.Printf("Enabling versioning on GCS bucket %s", config.remoteStateConfigGCS.Bucket)
+		bucketAttrs.VersioningEnabled = true
+	}
+
+	err := bucket.Create(ctx, projectID, bucketAttrs)
+	return errors.WithStackTrace(err)
+}
+
+// GCP is eventually consistent, so after creating a GCS bucket, this method can be used to wait until the information
+// about that GCS bucket has propagated everywhere.
+func WaitUntilGCSBucketExists(gcsClient *storage.Client, config *RemoteStateConfigGCS, terragruntOptions *options.TerragruntOptions) error {
+	terragruntOptions.Logger.Printf("Waiting for bucket %s to be created", config.Bucket)
+	for retries := 0; retries < MAX_RETRIES_WAITING_FOR_GCS_BUCKET; retries++ {
+		if DoesGCSBucketExist(gcsClient, config) {
+			terragruntOptions.Logger.Printf("GCS bucket %s created.", config.Bucket)
+			return nil
+		} else if retries < MAX_RETRIES_WAITING_FOR_GCS_BUCKET-1 {
+			terragruntOptions.Logger.Printf("GCS bucket %s has not been created yet. Sleeping for %s and will check again.", config.Bucket, SLEEP_BETWEEN_RETRIES_WAITING_FOR_GCS_BUCKET)
+			time.Sleep(SLEEP_BETWEEN_RETRIES_WAITING_FOR_GCS_BUCKET)
+		}
+	}
```

Comment:
Any reason not to use `retry.DoWithRetry` here?

Reply:
@brikis98 I guess because `DoesGCSBucketExist` returns a bool and not an error. I could change it to return an error and a bool or is there a cleaner way to do it in the retry func?